src/execution_plan/optimizations/optimizer.c

```diff
@@ -24,6 +24,9 @@ void optimizePlan(ExecutionPlan *plan) {
 	/* Remove redundant SCAN operations. */
 	reduceScans(plan);
 
+	/* Try to optimize cartesian product */
+	optimizeCartesianProduct(plan);
```

Comment:
It would be more efficient to call `applyJoin` before `optimizeCartesianProduct`, as the former may reduce the number of CPs with > 2 children.